builtin/credential/aws/path_role.go

```diff
@@ -319,6 +319,113 @@ func (b *backend) setRole(ctx context.Context, s logical.Storage, roleName strin
 	return nil
 }
 
+// initialize is used to initialize the AWS roles
+func (b *backend) initialize(ctx context.Context, req *logical.InitializationRequest) error {
+
+	// on standbys and DR secondaries we do not want to run any kind of upgrade logic
+	if b.System().ReplicationState().HasState(consts.ReplicationPerformanceStandby | consts.ReplicationDRSecondary) {
+		return nil
+	}
+
+	// Initialize only if we are either:
+	//   (1) A local mount.
+	//   (2) Are _NOT_ a replicated performance secondary
+	if b.System().LocalMount() || !b.System().ReplicationState().HasState(consts.ReplicationPerformanceSecondary) {
+
+		s := req.Storage
+
+		logger := b.Logger().Named("initialize")
+		logger.Info("starting initialization")
+
+		go func() {
+			// The vault will become unsealed while this goroutine is running,
+			// so we could see some role requests block until the lock is
+			// released.  However we'd rather see those requests block (and
+			// potentially start timing out) than allow a non-upgraded role to
+			// be fetched.
+			b.roleMutex.Lock()
+			defer b.roleMutex.Unlock()
+
+			upgraded, err := b.upgrade(ctx, s)
+			if err != nil {
+				logger.Error("error running initialization", "error", err)
+				return
+			}
+			if upgraded {
+				logger.Info("an upgrade was performed during initialization")
+			}
+
+			logger.Trace("initialization succeeded")
+		}()
+
+	}
+
+	return nil
+}
+
+// awsVersion stores info about the the latest aws version that we have
+// upgraded to.
+type awsVersion struct {
+	RoleVersion int `json:"role_verison"`
+}
+
+// upgrade aws version
+func (b *backend) upgrade(ctx context.Context, s logical.Storage) (bool, error) {
+
+	// find the persisted role version
+	roleVersion := -1
+	entry, err := s.Get(ctx, "config/version")
+	if err != nil {
+		return false, err
+	}
+	if entry != nil {
+		var rsv awsVersion
+		err = entry.DecodeJSON(&rsv)
+		if err != nil {
+			return false, err
+		}
+		roleVersion = rsv.RoleVersion
+	}
+
+	// for now, we only need to upgrade if the persisted role version
+	// has been superseded. This may change in the future.
+	needToUpgrade := roleVersion < currentRoleStorageVersion
+	if !needToUpgrade {
+		return false, nil
+	}
+
+	// Read all the role names.
+	roleNames, err := s.List(ctx, "role/")
+	if err != nil {
+		return false, err
+	}
+
+	// Upgrade the roles as necessary.
```

Comment:
Still missing the switch based mechanism from role upgrades. Looks good otherwise.